Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean

```diff
@@ -0,0 +1,269 @@
+    else
+      absorbLambdas lambdas' type args otherArgs
+
+/--
+Normalize an application of a constant with a Pi-type instance which distributes,
+by distributing over-applied arguments and/or distributing all lambdas.
+
+Note: we return a `List` since this can have multiple results.
+-/
+@[specialize] private def reduceAux (type inst : Expr) (args : List Expr) (lambdas : List FVarId)
+    (allArgs : Array Expr) (otherArgs : Array Expr) (arity numInstArgs : Nat)
+    (matchPiInst : Expr → Option Expr) (mk : Expr → List Expr → Array (Option Expr)) :
+    MetaM (List (Array (Option Expr) × List FVarId)) := do
+  let (type, args, idx) ← absorbArgs allArgs arity type inst args matchPiInst numInstArgs
+  match idx with
+  | some idx => return [(mk type args ++ allArgs[idx:].toArray.map some, lambdas)]
+  | none =>
+    (← absorbLambdas lambdas type args otherArgs).mapM fun (type, args, lambdas) =>
+      return (mk type args, lambdas)
+
+
+/-- Introduce new lambdas by η-expansion to reach the required number of arguments. -/
+@[specialize]
+private def etaExpand {α} (args : Array Expr) (type : Expr) (lambdas : List FVarId) (arity : Nat)
+    (k : (args : Array Expr) → List FVarId → (arity ≤ args.size) → MetaM α) : MetaM α  := do
+  if h : args.size < arity then
+    withLocalDeclD `_η type fun fvar =>
+      etaExpand (args.push fvar) type (fvar.fvarId! :: lambdas) arity k
+  else
+    k args lambdas (by omega)
+termination_by arity - args.size
+
+@[inline] private def app6? (fName : Name) (e : Expr) : Option (Expr × Expr × Expr) :=
+  if e.isAppOfArity fName 6 then
+    some (e.appFn!.appFn!.appArg!, e.appFn!.appArg!, e.appArg!)
+  else
+    none
+
+private def lambdaBody? (n : Nat) (e : Expr) : Option Expr :=
+match n, e with
+| 0, e => some e
+| n+1, .lam _ _ b _ => lambdaBody? n b
+| _, _ => none
+
+/-- Normalize an application if the head is `⁻¹` or `-`. -/
+private def reduceUnOp (n : Name) (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 2 then return none else some <$> do
+  let type := args[0]
+  let inst := args[1]
+  etaExpand args type lambdas 3 fun args lambdas _ => do
+    let arg := args[2]
+    reduceAux type inst [arg] lambdas args #[inst] 3 1
+      (lambdaBody? 2 >=> (·.app3? n) >=> fun (_, inst, a) =>
+        -- check for `fun a x => -a x` or `fun a x => (a x)⁻¹`
+        if (a == .app (.bvar 1) (.bvar 0)) then inst else none)
+      (fun
+        | type, [arg] => #[type, none, arg]
+        | _, _ => unreachable! )
+
+/-- Normalize an application if the head is `•` or `+ᵥ`. -/
+private def reduceHActOp (n : Name) (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
+  if h : args.size < 5 then return none else
+  let rec instH := match n with
+    | `HVAdd.hVAdd => `instHVAdd
+    | `HSMul.hSMul => `instHSMul
```

Comment:
Can you use ``` `` ``` to make this an error if the names change?

Reply:
I did that oroginally, but it requires to import more, and then it complains about an unnecessary import

Reply:
The unnecessary import can be silenced by running `lake exe shake --update`. Unless the import is particularly expensive (which CI will tell you), I think robustness matters more.